core/src/test/java/kafka/server/BrokerRegistrationRequestTest.java

```diff
@@ -0,0 +1,217 @@
+        NodeToControllerChannelManager channelManager,
+        AbstractRequest.Builder<T> reqBuilder,
+        int timeoutMs
+    ) throws Exception {
+        var responseFuture = new CompletableFuture<R>();
+        channelManager.sendRequest(reqBuilder, new ControllerRequestCompletionHandler() {
+            @Override
+            public void onTimeout() {
+                responseFuture.completeExceptionally(new TimeoutException());
+            }
+
+            @Override
+            public void onComplete(ClientResponse response) {
+                responseFuture.complete((R) response.responseBody());
+            }
+        });
+        return responseFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
+    }
+
+    public Errors registerBroker(
+        NodeToControllerChannelManager channelManager,
+        String clusterId,
+        int brokerId,
+        Long zkEpoch,
+        FeatureLevel featureLevelToSend
+    ) throws Exception {
+        var features = new BrokerRegistrationRequestData.FeatureCollection();
+
+        if (featureLevelToSend != null) {
+            features.add(new BrokerRegistrationRequestData.Feature()
+                    .setName(MetadataVersion.FEATURE_NAME)
+                    .setMinSupportedVersion(featureLevelToSend.min())
+                    .setMaxSupportedVersion(featureLevelToSend.max())
+            );
+        }
+        
+        Feature.PRODUCTION_FEATURES.stream()
+            .filter(feature -> !feature.featureName().equals(MetadataVersion.FEATURE_NAME))
+            .forEach(feature -> features.add(new BrokerRegistrationRequestData.Feature()
+                .setName(feature.featureName())
+                .setMinSupportedVersion(feature.minimumProduction())
+                .setMaxSupportedVersion(feature.latestTesting())));
+
+        var listener = new BrokerRegistrationRequestData.Listener()
+                .setName("EXTERNAL")
+                .setHost("example.com")
+                .setPort(8082)
+                .setSecurityProtocol(SecurityProtocol.PLAINTEXT.id);
+        
+        var req = new BrokerRegistrationRequestData()
+            .setBrokerId(brokerId)
+            .setLogDirs(List.of(Uuid.randomUuid()))
+            .setClusterId(clusterId)
+            .setIncarnationId(Uuid.randomUuid())
+            .setIsMigratingZkBroker(zkEpoch != null)
+            .setFeatures(features)
+            .setListeners(new BrokerRegistrationRequestData.ListenerCollection(List.of(listener).iterator()));
+
+        BrokerRegistrationResponse resp = this.sendAndReceive(
+            channelManager, 
+            new BrokerRegistrationRequest.Builder(req), 
+            30000
+        );
+        return Errors.forCode(resp.data().errorCode());
+    }
+
+    @ClusterTest(types = {Type.KRAFT}, controllers = 1, metadataVersion = MetadataVersion.IBP_3_3_IV3)
+    public void testRegisterZkWith33Controller(ClusterInstance clusterInstance) throws Exception {
```

Comment:
This test name is no longer accurate, since the RPCs are not all related to ZK. Perhaps you could split the scenarios into separate test cases?